### ISSUE TITLE
Implement IPFS streaming API

### ISSUE THREAD

**Issue by Jannis** [graphprotocol/graph-node#815]

**Comment by leodasvacas** [comment]
So, the idea here is to have an `ipfs.cat` that can be used for files that are too large to fit in memory. It is particulary useful with the independent feature of having a bootstrapping `init` function in mappings, which is run before any events are processed.

The signature for `catLines` could be `function catLines(hash: string, callback: string): void`, where `hash` is the ipfs address of an UTF8 file and `callback` is the name of a function with signature `function foo(line: string): void`. When `catLines` is called, we'll read the lines on the ipfs file stream and feed them to the callback. There is a limit on the maximum bytes lenght of a line, set through an env var. Once the file is done `catLines` returns. If an error is encountered, `catLines` traps, as any other of our host functions.

We currently call each event handler from a freshly instantiated `WasmiModule`, and we may do the same for the callback calls. We'd instantiate a module from within the host function of the "master" module, re-instantiating it for each call.

Does this sound like a plan?

**Comment by Jannis** [comment]
One additional comment: `function foo(foo: string): void` is nice; to support JSONLines files (ENS may end up using this), we also need to add a `json.fromString(data: string): JSONValue` host exported function, because currently we only have `json.fromBytes(data: Bytes): JSONValue`. Then ENS and others can do
```
function handleLin(data: string): void {
  let value = json.fromString(data)
  ...
}
```

**Comment by lutter** [comment]
Thanks for the comments! Just to be clear on the main limitation of this approach. I understand the scope of this issue to solely about files that can be ingested in one shot. In particular, resuming an ingestion that got half way through a file is not in scope for this. We'd rely on users to work within that limitation. For example, it's highly likely that using `catLines` with a very large file will run into trouble (losing connection to IPFS, index node getting shut down in the middle of ingestion) If we need more sophistication to make that easier on users, we'll address that in a separate issue/feature.

As for ingesting JSON: using `catLines` for this seems like a hack to me, since JSON has no notion of lines, and relying on a special format seems very brittle. I think it would be better to also offer a `catJson` function that reads a file (stream) and produces a stream of JSON values. It seems that serde's [StreamDeserializer](https://docs.serde.rs/serde_json/struct.StreamDeserializer.html) is an excellent fit for this. With that, we would expect the input file to contain a sequence of JSON values (not an array, just values one after the other like in the example in the docs page)

One final thought: should we make this interface a little more generic, like `function ipfsStream(hash: string, callback: string, flags: string[]) : someSortOfErrorIndicator` where `flags` could be used to indicate what the callback wants to be fed (`lines` or `json` for now) and maybe in the future also allow things like decompressing the content (say, with a `gzip` flag to indicate that the file is gzipped)

**Comment by lutter** [comment]
One more question: @leodasvacas, you mentioned that we might want to invoke the callback in a separate WasmiModule. But it seems to me that if users want to do any sort of processing on everything they got from `catLines`, it would be better to run the callback in the same module instance that called `catLines`. Otherwise, they would have to bounce things through the database, which can get expensive. Is there a security/separation issue that makes running in a separate module desirable?

**Comment by leodasvacas** [comment]
@lutter I agree with everything you said, your proposed interface is a big improvement! We were thinking of http://jsonlines.org/ but `StreamDeserializer` is just awesome, I didn't know that was a thing. Initially I was hoping we could somehow make this resumable, but I agree that will not be possible. @Jannis is that an acceptable limitation?

I still have one reasoning for separating modules and clearing state for the handler, and that's memory management. Our current memory management strategy is simply to never free memory and instead reset the module. I do not see a way to do better than this right now. Entities would not go through the database because the child modules should share the in-memory entity operations with the parent module.

This does raise an issue which I did not intially consider, which is that the in-memory operations will have to periodically be flushed to disk. But if we flush them to the database they must not be treated as regular entity data that appears in queries and is persisted across restarts. The only concern is to not keep a million entity operations in RAM, we could even just use the filesystem if that's better. But I'm sure you can think of good solutions here.

**Comment by lutter** [comment]
I agree with all that. It seems to me that no matter what we do, this will not be a good interface for bulk import of data - it'll be the right thing for pulling a relatively small amount of data from IPFS, but not for importing large amounts of it.

But bulk import might be a lot simpler than we thought: it would be reasonable to expect the data to be brought into a format that we define offline, so that we can pull it in at once, without running any user handlers. After all, this is about handling fixed, immutable data, and is therefore quite different from a block stream.

 I am thinking of something like `bulkImport(hash: string, entity: string, flags:string[])` where we'd expect `hash` to point to an IPFS file that contains a list of JSON objects, each of which represents an entity of type `entity` (defined in the subgraph's GraphQL schema, and their JSON representation follows that schema) The flags would let you control whether the import should replace all the entities of that type, or should only do an upsert of the ones mentioned in the file. For users, it shouldn't matter all that much whether we run the code that transforms entities, or whether they run that before publishing to IPFS; for extra credit, we could add a tool to check the format of the transformed data before users upload to IPFS. We can do this sort of bulk import much more efficiently than what the `ipfsStream` interface would allow (maybe even with a Postgres `COPY`, but definitely with a bulk insert; and we'd have a lot of latitude in how exactly the import happens).

What's not clear to me is how a bulk import would be triggered: would you have to deploy a new subgraph version to kick off a bulk import, or would you call that from a handler, based on data in an event?

If we want to push that even further, we could also go to a file format that's a little more inspired by git, where the `hash` points at a file that lists the hashes of the actual files to ingest, with some rules on how several files are combined into one logical file (e.g., if the same `ID` is mentioned in different files, the first one wins) That would make it quite nice for users to publish updates to existing files without having to republish everything again. And we could presumably figure out what changed, and only ingest the changes.

@leodasvacas and @Jannis: what do you think? And which interface do you think is more important to have now - the `ipfsStream` for smaller amounts of data or `bulkImport` for large amounts?

**Comment by Jannis** [comment]
I can think of two situations where we need this right now:

1. ENS: A 6GB+ file for hash -> name mappings.
2. Decentraland: A small file for district definitions (there are only a few districts).

A few other notes:

* Bulk import has to be triggered from the subgraph in some way (manifest or mapping).

* We will want IPFS data sources eventually, but until then, it is fine if the import is triggered from inside a mapping. E.g. in the ENS subgraph, it could be triggered when the first event for the root domain is processed. Worth double-checking though if that's easy to identify.

* We can't expect those files to contain read-to-use entities, so each of the values in these files has to go through a custom handler.

* I like the `ipfs.stream(hash: string, handler: string, flags: string[]): someSortOfErrorIndicator` proposal (note: I've renamed it to be in the `ipfs` namespace). `StreamDeserializer` seems to support a superset of JSON Lines, so that could be a great option for the implementation.

* Let's try implementing this without extra robustness first and see how well that works. It's likely that we'll have to implement something like resuming on startup (we may be able to get away with storing, for each such stream, the IPFS hash, the subgraph deployment and the "position").

* It's safe to assume that re-indexing the same file with the same handlers will generate the same entities, so it's also safe to flush entities to the store as soon as we receive them from the handlers.

* Let's keep it simple and only support single-file import per `ipfs.stream` call. If we want to add IPFS directory imports later, we can either add a `flag` for that or an `ipfs.streamDirectory` API or something in that direction.

**Comment by Jannis** [comment]
@leodasvacas Could you clarify the memory / flushing problem? I'm not sure I see any challenges there.

**Comment by leodasvacas** [comment]
@lutter I'm with @Jannis here in that we'll want to retain the flexibility of `ipfs.stream`. Also I don't see why you say that "this will not be a good interface for bulk import of data" and "We can do this sort of bulk import much more efficiently". I don't think resumability is of so much value, even for big files we should be able to handle them in under an hour. A simple handler should only account for a few milliseconds, pulling the bytes from ipfs should be the bottleneck, so skipping the handler might not be a huge performance gain.

@Jannis The challenge is that we should not write entities to the store after each line handler, because we transact entities to the store in the block glanurity, keeping the pending operations in memory (in the code, `entity_operations: Vec<EntityOperation>` in `EventHandlerContext`). So either the whole block is processed or it is entirely rolled back, keeping the store state consistent. But we do not want `entity_operations` to bloat our RAM usage when going through a huge ipfs file in a single block, so we should retain the same semantics but with the ability of flushing the operations to disk (a file or table or something) if necessary.

**Comment by Jannis** [comment]
@leodasvacas Ah, I was always thinking of the function as "spawning" an IPFS stream that is processed out of band, i.e., detached from the block stream. For one-shot bulk import scenarios like ENS and Decentraland the data isn't tied to a block, it's just that we "spawn" the processing from a block stream handler.

**Comment by Jannis** [comment]
That could be controlled with one of the flags perhaps.

**Comment by lutter** [comment]
@leodasvacas my comment about `ipfs.stream` not being a good interface for bulk import was based on your comments about when to commit entities into the store - doing it after each callback runs the risk of writing only parts of the file to the store (how would a user detect that? how do they fix it?) Doing it only after we've gone through the whole file runs the risk of accumulating a large (unbounded) number of entities in memory. In addition, because the callback has access to the store, we have very little understanding of what exactly happens in the callback, and how it interacts with the other entities already in the store. The performance comment was a reaction to your idea of invoking the callback in a separate module - I understood you to mean that each callback should be run in a freshly instantiated module, which IIRC takes ~100ms and would limit ingestion to ~ 36000 entities/hour.

@Jannis The reason I suggested that users can do their transformation offline was based on my understanding that at least for ENS, these files would be posted to IPFS just for use with The Graph anyway. From the user's point of view, the main difference between doing the transformation offline vs in a callback seems pretty small: it's whether they write a script that they run on their machine before posting to IPFS vs writing a callback that works in the context of `graph-node`. In my mind, IPFS import is quite different from handling a block stream in that IPFS files are fixed, and there is no evolving stream to follow.

In any event, there seems to be enough need for an `ipfs.stream` type interface, and I'll start there. The biggest question around that to me is still  when to store entities produced by the callback.  It would be easiest to expect users to do this the same way they do it in handlers now, by calling `save` explicitly, and setting the expectation that no in-memory state is preserved between invocations of the callback (basically by using a separate module to call the callback as @leodasvacas suggested)

**Comment by Jannis** [comment]
It's true that the two examples I've mentioned are custom files for The Graph. But this should still be flexible enough to perform transformations. Parallelization may be an option here to speed things up. If necessary. 

And yes, entities would be created and saved just like in other handlers. I would spawn a new Tokio task for every `ipfs.stream` invocation and store entities as soon as they are returned. Or maybe batch them using a batching size or time window but really, those optimizations. As long as we know it are possible with the design, they should always come at the end if deemed necessary.

**Comment by leodasvacas** [comment]
@Jannis doing this out of band is very similar to dynamically starting an IPFS data source, which brings up the discussion related to multiple data sources that we want to avoid for now. So can we do the semantics of the operations being within the block?

@lutter Ah I can see that I didn't communicate very well what I meant on these details, want to do a call on this so we can discuss looking at the code?

**Comment by Jannis** [comment]
Current idea after discussing it further offline:

- When `ipfs.stream` is called from a handler, process the entire stream in a blocking way before returning the the original handler. Through `.save()` all entity operations will be added to the existing vector of operations.
- We suspend the original handler's timeout while processing the stream and instead set up new timeouts for individual IPFS stream handler invocations.
- If an error happens during processing, we fail the IPFS stream and the subgraph.
- Every invocation of an IPFS stream handler happens in a new `WasmiModule`.
- We may need to limit the file size that we can iterate over.
- We may realize that this solution is not sufficient for _huge_ file imports. In that case, an `initHandler` per data source or splitting up e.g. the ENS file into smaller files could be options going forward.

**Comment by lutter** [comment]
I'll call what we called `ipfs.stream` in the above `ipfs.map` from now on. The signature of the function in AssemblyScript will be `ipfs.map(hash: string, callback: string, flags: string[]): void`

Tasks:

- [ ] write user-level docs for `ipfs.map`
- [ ] implement the `ipfs.map` version that will process a stream of JSON values and pass them to the `callback` following the design in the previous comment

**Comment by Jannis** [comment]
@lutter Should we split out the JSON stream implementation from the `ipfs.map` functionality in the second task? I think it makes sense to first implement the JSON stream based on an IPFS cat call/request and then wire that up with `ipfs.map`.

No other objections though!

**Comment by lutter** [comment]
@Jannis I added the task for graph-ts. I have the JSON stream working (albeit along jsonlines, since it turns out wiring up the stream you get from `IpfsClient.cat` to `serde_json::StreamDeserializer` is not really possible (that wants a `std::io::Read` and there's no good way to get there, which is super annoying because somewhere under all the abstractions is a socket, but you can't get to it) 

I should be able to wire this up as a host export tomorrow, which would give us the basic `ipfs.map` I described in the second task. So probably not worth splitting out further.

**Comment by lutter** [comment]
The `graph-ts` side of things is in [this PR](https://github.com/graphprotocol/graph-ts/pull/45)

**Comment by leodasvacas** [comment]
Though there are improvements to be made, the core of this is done, closing.
